datafusion/expr/src/udaf.rs

```diff
@@ -354,6 +360,24 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
     fn aliases(&self) -> &[String] {
         &[]
     }
+
+    /// Construct an expression that calculates the aggregate in reverse.
+    /// Typically the "reverse" expression is itself (e.g. SUM, COUNT).
+    /// For aggregates that do not support calculation in reverse,
+    /// returns None (which is the default value).
+    fn reverse_expr(&self) -> ReversedExpr {
+        ReversedExpr::NotSupported
+    }
+}
+
+#[derive(Debug)]
+pub enum ReversedExpr {
+    /// The expression is the same as the original expression, like SUM, COUNT
+    Identical,
+    /// The expression does not support reverse calculation, like ArrayAgg
+    NotSupported,
```

Comment:
I spent some time trying to understand this -- is the reason that ArrayAgg doesn't support reversed calculations that the array elements would be in a different order?

Reply:
Yes, because ArrayAgg has no ordering info. 

You can also take a look at this comment https://github.com/apache/datafusion/issues/9972#issuecomment-2055582944

Although I found that the counter-example might not be correct after a few days, `ARRAY_AGG(b ORDER BY c DESC)` is OrderSensitiveArrayAgg, so it is possible to produce reverse expr, but it still makes sense to me that order insensitive ArrayAgg (which is ArrayAgg) does not support `reverse_expr`. cc @mustafasrepo